Add some new methods to QgsBox3d

ptitjano opened the pull request:
## Description

This adds some new functionalities to `QgsBox3d`, mainly:
- `toString` method
- `is3D` method
- `isEmpty` method
- `isNull` method
- `combineWith` method
- simplify some logic if the box is 2d

cc @benoitdm-oslandia

benoitdm-oslandia committed:
qgsbox3d: Add 3d optional parameters to the QgsRectangle constructor

benoitdm-oslandia committed:
qgsbox3d: Add a normalize step parameter to the constructors

This allows to choose if a normalization step needs to be applied. It
is set to True by default to allow backward compatibility with the
api.

benoitdm-oslandia committed:
qgsbox3d: Treat a box with Z nan values as a 2d box

benoitdm-oslandia committed:
qgsbox3d: Add isNull method

This has is the following logic:

- QgSBox3d().isNull() == True
- QgSBox3d().isNull(0, 0, 0, 0, 0, 0) == False
- QgSBox3d().isNull(1, 1, 1, 1, 1, 1) == False

benoitdm-oslandia committed:
qgsbox3d: Add isEmpty method

This has is the following logic:

- QgSBox3d().isEmpty() == True
- QgSBox3d().isEmpty(0, 0, 0, 0, 0, 0) == True
- QgSBox3d().isEmpty(1, 1, 1, 1, 1, 1) == True

benoitdm-oslandia committed:
qgsbox3d: Add support for toString method

benoitdm-oslandia committed:
qgsbox3d: Add is3D method

benoitdm-oslandia committed:
qgsbox3d: Simplify some methods in the 2d case

benoitdm-oslandia committed:
qgsbox3d: Add a new contains method based on x,y,z coordinates

benoitdm-oslandia committed:
qgsbox3d: Add combineWith methods

benoitdm-oslandia committed:
qgsbox3d: Add a setMinimal method

nyalldawson commented:
Can you take a look at https://github.com/qgis/QGIS/pull/53323/files and adapt the definition of empty/null here to that PR? Otherwise we'll run into the same issue here where an empty box at 0,0,0 is incorrectly considered null.

benoitdm-oslandia committed:
qgsbox3d: Use nan for default values

This will be necessary to properly define `isNull` method in the next
commit.
This way, it behaves the same as `QgsPoint`.

Co-Authored by Jean Felder

nyalldawson commented on src/core/geometry/qgsbox3d.cpp:
```diff
@@ -164,3 +253,48 @@ void QgsBox3d::scale( double scaleFactor, double centerX, double centerY, double
+
+bool QgsBox3d::isNull() const
+{
+  return qgsDoubleNear( mBounds2d.xMinimum(), std::numeric_limits<double>::max() ) && qgsDoubleNear( mBounds2d.yMinimum(), std::numeric_limits<double>::max() ) && ( qgsDoubleNear( mZmin, std::numeric_limits<double>::max() ) || std::isnan( mZmin ) )
```
```suggestion
  return mBounds2d.xMinimum() == std::numeric_limits<double>::max() && mBounds2d.xMaximum() == std::numeric_limits<double>::lowest() mBounds2d.yMinimum() == std::numeric_limits<double>::max() && mBounds2d.yMaximum() == std::numeric_limits<double>::lowest() && ( ( mZmin == std::numeric_limits<double>::max() && mZmax == std::numeric_limits<double>::lowest() ) || ( std::isnan( mZmin ) && std::isnan( mZmax) ) )
```

ptitjano replied:
Done.

ptitjano commented:
> Can you take a look at https://github.com/qgis/QGIS/pull/53323/files and adapt the definition of empty/null here to that PR? Otherwise we'll run into the same issue here where an empty box at 0,0,0 is incorrectly considered null.

I have updated the PR, this should now behave as #53323. Especially `QgsBox3d(0, 0, 0, 0, 0)` is not considered null.

ptitjano commented on python/core/auto_generated/geometry/qgsbox3d.sip.in:
```diff
@@ -152,6 +154,14 @@ Returns the maximum z value.
+
+    void setMinimal() /HoldGIL/;
+%Docstring
+Set a rectangle so that min corner is at max
```
Fixed.

ptitjano commented on python/core/auto_generated/geometry/qgsbox3d.sip.in:
```diff
@@ -218,6 +236,30 @@ Returns ``True`` when box contains a ``point``.
+Returns ``True`` when box contains a point (``x``, ``y``, ``z``).
+
+If the point is a 2D point (no z-coordinate), then the containment test
+will be performed on the x/y extent of the box only.
```
Fixed.

lbartoletti commented on src/core/geometry/qgsbox3d.cpp:
```diff
@@ -86,47 +112,110 @@ QgsBox3d QgsBox3d::intersect( const QgsBox3d &other ) const
+
+bool QgsBox3d::is3D() const
+{
+  return !is2d() && !isNull();
```
Isn't possible to have a case where `( mZmin > mZmax )` and have to consider the box as an "inversed" 3D Box?

ptitjano replied:
I think, it's safe at the moment to assume that It returns `FALSE` if the box is not normalized. I have added a sentence about it in the function description.

lbartoletti commented on src/core/geometry/qgsbox3d.cpp:
```diff
@@ -86,47 +112,110 @@ QgsBox3d QgsBox3d::intersect( const QgsBox3d &other ) const
 bool QgsBox3d::contains( const QgsPoint &p ) const
 {
-  if ( !mBounds2d.contains( p.x(), p.y() ) )
+  if ( p.is3D() && is3D() )
```
I think it's safe to remove the `p.is3D()` here since you test if it's a nan in `contains(double, double, double)`.

lbartoletti commented on src/core/geometry/qgsbox3d.cpp:
```diff
@@ -86,47 +112,110 @@ QgsBox3d QgsBox3d::intersect( const QgsBox3d &other ) const
-  return sqrt( dx * dx + dy * dy + dz * dz );
+  if ( is2d() || std::isnan( point.z() ) )
+  {
+    return sqrt( dx * dx + dy * dy );
```
One mayuse [hypot](https://en.cppreference.com/w/cpp/numeric/math/hypot)

ptitjano commented:
Update: fixed test `PyQgsOapifProvider`
